src/librustc_mir/util/pretty.rs

```diff
@@ -534,6 +538,197 @@ pub fn write_mir_intro<'tcx>(
+/// the raw bytes where applicable.
+/// The byte format is similar to how hex editors print bytes. Each line starts with the address of
+/// the start of the line, followed by all bytes in hex format (space separated).
+/// If the allocation is small enough to fit into a single line, no start address is given.
+///
+/// The `prefix` argument allows callers to add an arbitrary prefix before each line (even if there
+/// is only one line). Note that your prefix should contain a trailing space as the lines are
+/// printed directly after it.
+pub fn write_allocation<Tag, Extra>(
+    tcx: TyCtxt<'tcx>,
+    alloc: &Allocation<Tag, Extra>,
+    w: &mut dyn Write,
+    prefix: &str,
+) -> io::Result<()> {
+    const BYTES_PER_LINE: Size = Size::from_bytes(16);
+    let num_lines = alloc.size.bytes().saturating_sub(BYTES_PER_LINE.bytes()) as usize;
+    // Number of chars needed to represent all line numbers.
+    let pos_width = format!("{:x}", alloc.size.bytes()).len();
+
+    if num_lines > 0 {
+        write!(w, "{}0x{:02$x} │ ", prefix, 0, pos_width)?;
+    } else {
+        write!(w, "{}", prefix)?;
+    }
+
+    let mut i = Size::ZERO;
+    let mut line_start = Size::ZERO;
+
+    // Prints the line start address.
+    let newline = |w: &mut dyn Write, line_start: &mut Size| {
+        *line_start += BYTES_PER_LINE;
+        write!(w, "\n{}0x{:02$x} │ ", prefix, line_start.bytes(), pos_width)
+    };
+
+    let ptr_size = tcx.data_layout.pointer_size;
+
+    while i < alloc.size {
+        // The line start already has a space. While we could remove that space from the line start
+        // printing and unconditionally print a space here, that would cause the single-line case
+        // to have a single space before it, which looks weird.
+        if i != line_start {
+            write!(w, " ")?;
+        }
+        if let Some(&(_, target_id)) = alloc.relocations().get(&i) {
+            // Memory with a relocation must be defined
+            let j = i.bytes() as usize;
+            let offset = alloc
+                .inspect_with_undef_and_ptr_outside_interpreter(j..j + ptr_size.bytes() as usize);
+            let offset = read_target_uint(tcx.data_layout.endian, offset).unwrap();
+            // this `as usize` is fine, since we can't print more chars than `usize::MAX`
+            let relocation_width = |bytes| bytes as usize * 3;
+            let mut target = format!("alloc{}+{}", target_id, offset);
+            if (i - line_start) + ptr_size > BYTES_PER_LINE {
+                // This branch handles the situation where a relocation starts in the current line
+                // but ends in the next one.
+                let remainder = BYTES_PER_LINE - (i - line_start);
+                let overflow = ptr_size - remainder;
+                let remainder_width = relocation_width(remainder.bytes()) - 2;
+                if target.len() > remainder_width {
+                    write!(target, " ({} bytes)", ptr_size.bytes()).unwrap();
```

Comment:
Would "N ptr bytes" help resolve your concern about this looking too much like the allocation size, @oli-obk?